### PR TITLE
CA1810 Initialize reference type static fields inline

### PR DESCRIPTION
**Pull request by elachlan**
Relates to #7174

### CODE REVIEW

File: src/Build/Evaluation/ProjectRootElementCache.cs [hunk 1]
```diff
@@ -69,20 +69,21 @@ internal class ProjectRootElementCache : ProjectRootElementCacheBase
         /// If this number is increased much higher, the datastructure may
         /// need to be changed from a linked list, since it's currently O(n).
         /// </remarks>
-        private static readonly int s_maximumStrongCacheSize = 200;
+        private static readonly int s_maximumStrongCacheSize = !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("MSBUILDPROJECTROOTELEMENTCACHESIZE")) ?
+            Convert.ToInt32(Environment.GetEnvironmentVariable("MSBUILDPROJECTROOTELEMENTCACHESIZE"), NumberFormatInfo.InvariantInfo)  : 200;
```
Comment (Forgind) [inline src/Build/Evaluation/ProjectRootElementCache.cs:73]: I'm tempted to propose:
```suggestion
        private static readonly int s_maximumStrongCacheSize = !string.IsNullOrEmpty((string s = Environment.GetEnvironmentVariable("MSBUILDPROJECTROOTELEMENTCACHESIZE"))) ?
            Convert.ToInt32(s, NumberFormatInfo.InvariantInfo) : 200;
```

Honestly not sure which I like better. I like inlining most of these, but this one is messy.

Comment (Forgind) [reply src/Build/Evaluation/ProjectRootElementCache.cs:73]: Actually, Environment.GetEnvironmentVariable returns null if it isn't defined; how about:
```suggestion
        private static readonly int s_maximumStrongCacheSize =
            Convert.ToInt32(Environment.GetEnvironmentVariable("MSBUILDPROJECTROOTELEMENTCACHESIZE") ?? "200", NumberFormatInfo.InvariantInfo);
```
?

File: src/MSBuild/XMake.cs [hunk 2]
```diff
@@ -124,7 +124,9 @@ public enum ExitType
         /// <summary>
         /// Static constructor
         /// </summary>
+#pragma warning disable CA1810 // Initialize reference type static fields inline
```
Comment (Forgind) [inline src/MSBuild/XMake.cs:127]: Why is this one problematic?

File: src/Shared/FrameworkLocationHelper.cs [hunk 3]
```diff
@@ -374,16 +374,10 @@ internal static class FrameworkLocationHelper
...
+        private static readonly IReadOnlyDictionary<Version, VisualStudioSpec> s_visualStudioSpecDict = s_visualStudioSpecs.ToDictionary(spec => spec.Version);
 
-#endregion // Static member variables
-
-        static FrameworkLocationHelper()
-        {
-            s_dotNetFrameworkSpecDict = s_dotNetFrameworkSpecs.ToDictionary(spec => spec.Version);
-            s_visualStudioSpecDict = s_visualStudioSpecs.ToDictionary(spec => spec.Version);
-        }
+        #endregion // Static member variables
```
Comment (Forgind) [inline src/Shared/FrameworkLocationHelper.cs:380]: nit:
unindent?

File: src/Build/Evaluation/ProjectRootElementCache.cs (hunk 1 above)
Comment (elachlan) [reply src/Build/Evaluation/ProjectRootElementCache.cs:73]: I also thought that, but what if the environment variable is an empty string value?

Comment (Forgind) [reply src/Build/Evaluation/ProjectRootElementCache.cs:73]: On Windows, empty strings are undefined for environment variables, so that shouldn't be an issue. Linux does differentiate between unset environment variables and empty ones, but ultimately, it shouldn't matter. This is for debugging/testing only and isn't published anywhere. Anyone who relies on this _and_ misuses it deserves to have a broken build.

(That's also why we don't verify that it's really an integer before we try to coerce it into one.)

**Review by Forgind — APPROVED** [review]

File: src/MSBuild/XMake.cs (hunk 2 above)
Comment (elachlan) [reply src/MSBuild/XMake.cs:127]: It is initializing static fields here. But there is a big block comment:
```
////////////////////////////////////////////////////////////////////////////////
//  Only initialize static fields here, not inline!                           //
//  This forces the type to initialize in this static constructor and thus    //
//  any configuration file exceptions can be caught here.                     //
////////////////////////////////////////////////////////////////////////////////
```
So I put a suppression in.

File: src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupLoggingHelper.cs [hunk 4]
```diff
@@ -41,7 +41,9 @@ internal static class ItemGroupLoggingHelper
         /// to materialize the Message as that's a declaration assembly. We inject the logic
         /// here.
         /// </summary>
+#pragma warning disable CA1810 // Initialize reference type static fields inline
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupLoggingHelper.cs:44]: Actually, I don't think this constructor is ever used; can we just delete it?

File: src/MSBuild/XMake.cs (hunk 2 above)
Comment (Forgind) [reply src/MSBuild/XMake.cs:127]: Same for this one

File: src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupLoggingHelper.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (elachlan) [reply src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupLoggingHelper.cs:44]: I think it is a false positive. Since the variables that it is assigning to are not members of the class. But are static members of other static classes.

Moving those inline would change the logic.

[the remark this answers is not in the document]
Comment (Forgind) [reply src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupLoggingHelper.cs:44]: Gotcha. Sounds good.

[the remark this answers is not in the document]
Comment (jmarolf) [reply src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupLoggingHelper.cs:44]: @elachlan can you please file an issue on https://github.com/dotnet/roslyn-analyzers about this false positive?

**Comment by elachlan** [comment]
There are still 3 occurrences that I believe are false positives. I can add a suppression, but I think if this takes a while to merge it might be fixed in the analyzer by then.

File: src/Tasks/CultureInfoCache.cs [hunk 5]
```diff
@@ -19,35 +19,36 @@ namespace Microsoft.Build.Tasks
...
-            ValidCultureNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
-
+            HashSet<string> validCultureNames = new(StringComparer.OrdinalIgnoreCase);
 #if !FEATURE_CULTUREINFO_GETCULTURES
             if (!AssemblyUtilities.CultureInfoHasGetCultures())
             {
-                ValidCultureNames = HardcodedCultureNames;
-                return;
+                validCultureNames = HardcodedCultureNames;
+                return validCultureNames;
```
Comment (Forgind) [inline src/Tasks/CultureInfoCache.cs:31]: nit:
```suggestion
                return HardcodedCultureNames;
```

And move this to the top of the function?

File: src/Tasks/GenerateResource.cs [hunk 6]
```diff
@@ -923,7 +900,7 @@ public override bool Execute()
         }
 
 #if FEATURE_COM_INTEROP
-        private static bool allowMOTW;
+        private static readonly bool AllowMOTW = !NativeMethodsShared.IsWindows || (Registry.GetValue(@"HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\.NETFramework\SDK", "AllowProcessOfUntrustedResourceFiles", null) as string ?? string.Empty).Equals("true", StringComparison.OrdinalIgnoreCase);
```
Comment (Forgind) [inline src/Tasks/GenerateResource.cs:903]:
```suggestion
        private static readonly bool AllowMOTW = !NativeMethodsShared.IsWindows || (Registry.GetValue(@"HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\.NETFramework\SDK", "AllowProcessOfUntrustedResourceFiles", null) is string allowUntrustedFiles && allowUntrustedFiles.Equals("true", StringComparison.OrdinalIgnoreCase));
```
?

File: src/Tasks/GetFrameworkPath.cs [hunk 7]
```diff
@@ -52,23 +32,22 @@ public override bool Execute()
...
+        private static readonly Lazy<string> s_version20Path = new Lazy<string>(() => ToolLocationHelper.GetPathToDotNetFramework(TargetDotNetFrameworkVersion.Version20));
+        private static readonly Lazy<string> s_version30Path = new Lazy<string>(() => ToolLocationHelper.GetPathToDotNetFramework(TargetDotNetFrameworkVersion.Version30));
+        private static readonly Lazy<string> s_version35Path = new Lazy<string>(() => ToolLocationHelper.GetPathToDotNetFramework(TargetDotNetFrameworkVersion.Version35));
+        private static readonly Lazy<string> s_version40Path = new Lazy<string>(() => ToolLocationHelper.GetPathToDotNetFramework(TargetDotNetFrameworkVersion.Version40));
+        private static readonly Lazy<string> s_version45Path = new Lazy<string>(() => ToolLocationHelper.GetPathToDotNetFramework(TargetDotNetFrameworkVersion.Version45));
+        private static readonly Lazy<string> s_version451Path=new Lazy<string>(() => ToolLocationHelper.GetPathToDotNetFramework(TargetDotNetFrameworkVersion.Version451));
```
Comment (Forgind) [inline src/Tasks/GetFrameworkPath.cs:42]: tiny nit:
```suggestion
        private static readonly Lazy<string> s_version451Path = new Lazy<string>(() => ToolLocationHelper.GetPathToDotNetFramework(TargetDotNetFrameworkVersion.Version451));
```

**Review by Forgind — APPROVED** [review]
Looks good! I'm assuming the false positive problem isn't about to be resolved, though we also shouldn't be merging right now because our last insertion failed for some reason, so we have a little time.

**Comment by elachlan** [comment]
> Looks good! I'm assuming the false positive problem isn't about to be resolved, though we also shouldn't be merging right now because our last insertion failed for some reason, so we have a little time.

There is only one false positive and it has a suppression now. The other two were me not fully reading the analyzer article and understanding what it wanted me to do. Hence the fixes I just submitted.

File: src/Build/Evaluation/ProjectRootElementCache.cs [hunk 8]
```diff
@@ -71,22 +71,21 @@ internal class ProjectRootElementCache : ProjectRootElementCacheBase
         /// If this number is increased much higher, the datastructure may
         /// need to be changed from a linked list, since it's currently O(n).
         /// </remarks>
-#pragma warning disable CA1802 // Use literals where appropriate
-        private static readonly int s_maximumStrongCacheSize = 200;
-#pragma warning restore CA1802 // Use literals where appropriate
+        private static readonly int s_maximumStrongCacheSize =
+            Convert.ToInt32(Environment.GetEnvironmentVariable("MSBUILDPROJECTROOTELEMENTCACHESIZE") ?? "200", NumberFormatInfo.InvariantInfo);
```
Comment (ladipro) [inline src/Build/Evaluation/ProjectRootElementCache.cs:75]: Can you please add a new private static method and:

```suggestion
        private static readonly int s_maximumStrongCacheSize = InitializeMaximumStrongCacheSize();
```

The initialization expression you wrote makes us parse the string `"200"` which is suboptimal.

File: src/MSBuild/XMake.cs (hunk 2 above)
Comment (ladipro) [reply src/MSBuild/XMake.cs:129]: Static ctors are invoked automatically by the runtime, yes. `MSBuildApp` is the entry point so it's not clear why we wouldn't want to do this work at the top of `Main` instead. It should be equivalent unless someone directly calls into other `MSBuildApp` methods (unlikely).

File: src/Shared/ExceptionHandling.cs [hunk 9]
```diff
@@ -34,12 +34,7 @@ namespace Microsoft.Build.Shared
     /// </summary>
     internal static class ExceptionHandling
     {
-        private static readonly string s_debugDumpPath;
-
-        static ExceptionHandling()
-        {
-            s_debugDumpPath = GetDebugDumpPath();
-        }
+        private static readonly string s_debugDumpPath = GetDebugDumpPath();
```
Comment (ladipro) [inline src/Shared/ExceptionHandling.cs:37]: This change may be counter-productive. Presumably the debug dump path is needed only if there actually was an exception. With this change we may be initializing it eagerly which is wasted work when no exception is thrown. We can afford to pay the small cost of precise (i.e. non-`beforefieldinit`) initialization on error paths.

Can you please perform a simple test to see if `GetDebugDumpPath` runs without your changes when the build succeeds?

File: src/Shared/FrameworkLocationHelper.cs [hunk 10]
```diff
@@ -376,17 +376,11 @@ internal static class FrameworkLocationHelper
        };
 #endif // FEATURE_WIN32_REGISTRY
 
-        private static readonly IReadOnlyDictionary<Version, DotNetFrameworkSpec> s_dotNetFrameworkSpecDict;
-        private static readonly IReadOnlyDictionary<Version, VisualStudioSpec> s_visualStudioSpecDict;
+        private static readonly IReadOnlyDictionary<Version, DotNetFrameworkSpec> s_dotNetFrameworkSpecDict = s_dotNetFrameworkSpecs.ToDictionary(spec => spec.Version);
+        private static readonly IReadOnlyDictionary<Version, VisualStudioSpec> s_visualStudioSpecDict = s_visualStudioSpecs.ToDictionary(spec => spec.Version);
```
Comment (ladipro) [inline src/Shared/FrameworkLocationHelper.cs:380]: This looks like expensive work. Making it run with the `beforefieldinit` semantics would be OK only if we're sure it's needed everywhere the type `FrameworkLocationHelper` is mentioned (not necessarily actually used).

**Review by ladipro — CHANGES_REQUESTED** [review]
Please note the tradeoffs being made when switching from an explicit static ctor to inline static field initializion. As the analyzer suggests, with `beforefieldinit` the JITted code may be slightly faster because it doesn't need to execute init checks. It may however lead to unwanted eager initialization. We should be especially careful when dealing with expensive initialization code. If there are scenarios where MSBuild "uses" the type but does not really call it or access its fields, then expensive initialization should be left in static ctors.

File: src/Shared/FrameworkLocationHelper.cs (hunk 10 above)
Comment (ladipro) [reply src/Shared/FrameworkLocationHelper.cs:380]: Apologies for the inconsistent use of the word "use".
- "Mention" here means the runtime sees the type because it is for example compiling a method referencing it. Making a method call in the code being compiled, accessing a field, a type check, ..., all these count.
- "Use" means an instance is actually created, a static method is called, or a static field is accessed. Actually as in at run-time.

File: src/Build/Evaluation/ProjectRootElementCache.cs (hunk 8 above)
Comment (ladipro) [reply src/Build/Evaluation/ProjectRootElementCache.cs:75]: That works too!

File: src/Shared/ExceptionHandling.cs (hunk 9 above)
Comment (Forgind) [reply src/Shared/ExceptionHandling.cs:37]: I feel like you teach me something new about C# every week 😁

When would static field be initialized without the static constructor being called? ["If static field variable initializers are present in the class of the static constructor, they're executed in the textual order in which they appear in the class declaration. The initializers run immediately prior to the execution of the static constructor."](https://docs.microsoft.com/dotnet/csharp/programming-guide/classes-and-structs/static-constructors) made it sound like they always happen together.

Comment (ladipro) [reply src/Shared/ExceptionHandling.cs:37]: They do happen together. Field initializers become part of the static ctor at IL level. It's just that depending on whether the class does or doesn't have a static ctor declared in C#, the code may run a different time. Let me link an older article explaining the difference: https://csharpindepth.com/articles/BeforeFieldInit

File: src/Shared/FrameworkLocationHelper.cs (hunk 10 above)
Comment (elachlan) [reply src/Shared/FrameworkLocationHelper.cs:380]: I can say with certainty that not all usages of `FrameworkLocationHelper` depend on `s_dotNetFrameworkSpecDict` or `s_visualStudioSpecDict`. 
Example:
https://github.com/dotnet/msbuild/blob/4bdda5bd1e70d2398badb79b92d20d27b840eaab/src/Shared/FrameworkLocationHelper.cs#L135

Does that mean `s_dotNetFrameworkSpecDict` and `s_visualStudioSpecDict` should keep their initialisation in the static constructor?

File: src/Build/Evaluation/ProjectRootElementCache.cs (hunk 8 above)
Comment (elachlan) [reply src/Build/Evaluation/ProjectRootElementCache.cs]: Done.

File: src/Build/Evaluation/ProjectRootElementCache.cs [hunk 11]
```diff
@@ -123,24 +122,6 @@ internal class ProjectRootElementCache : ProjectRootElementCacheBase
         /// </summary>
         private Object _locker = new Object();
```
Comment (elachlan) [inline src/Build/Evaluation/ProjectRootElementCache.cs:123]: I think this is a singleton and therefore will need to have a static constructor (even if it is empty). see https://csharpindepth.com/articles/Singleton

File: src/Shared/ExceptionHandling.cs (hunk 9 above)
Comment (elachlan) [reply src/Shared/ExceptionHandling.cs:37]: @ladipro I tested with a small .net framework project. It didn't hit `GetDebugDumpPath` when when the build succeeds before or after the change.

File: src/Shared/MSBuildNameIgnoreCaseComparer.cs [hunk 12]
```diff
@@ -22,16 +22,7 @@ internal class MSBuildNameIgnoreCaseComparer : IConstrainedEqualityComparer<stri
         /// <summary>
         /// The processor architecture on which we are running, but default it will be x86
         /// </summary>
-        private static readonly NativeMethodsShared.ProcessorArchitectures s_runningProcessorArchitecture;
-
-        /// <summary>
-        /// We need a static constructor to retrieve the running ProcessorArchitecture that way we can
```
Comment (elachlan) [inline src/Shared/MSBuildNameIgnoreCaseComparer.cs:28]: `We need a static constructor` did I overlook this? Is the static constructor needed to avoid issues on IA64?

Comment (ladipro) [reply src/Shared/MSBuildNameIgnoreCaseComparer.cs:28]: I don't think the comment was specifically about needing an explicit static ctor. Switching to inline initialization should be fine.

File: src/Shared/FrameworkLocationHelper.cs (hunk 10 above)
Comment (ladipro) [reply src/Shared/FrameworkLocationHelper.cs:380]: Confirming that there are scenarios where we run static initialization but end up not needing `s_dotNetFrameworkSpecDict` or `s_visualStudioSpecDict`. Due to the non-trivial cost of initializing static fields on this class I would discourage making the change in this PR to not risk regressions. Instead, please consider switching to an explicit lazy initialization of this data.

I see the static ctor run at this callstack, for example:

```
004fe734 04dcb6b8 Microsoft.Build.Shared.FrameworkLocationHelper..cctor()
004ff0ac 04dc823e Microsoft.Build.Internal.Utilities.GetEnvironmentProperties()
004ff0f0 04dc7c64 Microsoft.Build.Evaluation.ProjectCollection.get_EnvironmentProperties()
004ff124 04dc7bd0 Microsoft.Build.Evaluation.ProjectCollection.InitializeToolsetCollection(Microsoft.Build.Evaluation.ToolsetRegistryReader, Microsoft.Build.Evaluation.ToolsetConfigurationReader)
004ff13c 0248e676 Microsoft.Build.Evaluation.ProjectCollection..ctor(System.Collections.Generic.IDictionary`2, System.Collections.Generic.IEnumerable`1, System.Collections.Generic.IEnumerable`1, Microsoft.Build.Evaluation.ToolsetDefinitionLocations, Int32, Boolean, Boolean)
004ff190 0248cb74 Microsoft.Build.CommandLine.MSBuildApp.BuildProject(System.String, System.String[], System.String, System.Collections.Generic.Dictionary`2, System.Collections.Generic.Dictionary`2, Microsoft.Build.Framework.ILogger[], Microsoft.Build.Framework.LoggerVerbosity, Microsoft.Build.CommandLine.DistributedLoggerRecord[], Boolean, System.String, Int32, Boolean, System.IO.TextWriter, System.IO.TextWriter, Boolean, System.Collections.Generic.ISet`1, System.Collections.Generic.ISet`1, Boolean, Microsoft.Build.Logging.ProfilerLogger, Boolean, Boolean, Boolean, Microsoft.Build.Graph.GraphBuildOptions, Boolean, System.String[], System.String)
004ff2e0 02481e9f Microsoft.Build.CommandLine.MSBuildApp.Execute(System.String)
004ff3a4 024810e3 Microsoft.Build.CommandLine.MSBuildApp.Main()
```

where `GetEnvironmentProperties` accesses only the `programFiles32` and `programFiles64` fields.

This is further exacerbated by compiling `FrameworkLocationHelper` into multiple assemblies so in a typical build the static initialization runs twice per process, once for the copy in `Microsoft.Build`, once for the copy in `Microsoft.Build.Utilities.Core`.

In summary, this code is in need of larger refactoring. In this PR I would err on the safer side and leave it alone.

File: src/Shared/ExceptionHandling.cs (hunk 9 above)
Comment (ladipro) [reply src/Shared/ExceptionHandling.cs:37]: Thank you! Let's go with the change then. There's still a possibility that we will initialize more eagerly - depending on the runtime and the exact code path - plus the initialization has a side-effect of creating a directory (see the `DebugUtils` static ctor) but it's under a debug switch so no biggie.

File: src/Shared/FrameworkLocationHelper.cs (hunk 10 above)
Comment (elachlan) [reply src/Shared/FrameworkLocationHelper.cs]: I've changed it to be lazily initialized.

File: src/Shared/FrameworkLocationHelper.cs [hunk 13]
```diff
@@ -145,7 +145,7 @@ internal static class FrameworkLocationHelper
         /// <summary>
         /// List the supported .net versions.
         /// </summary>
-        private static readonly DotNetFrameworkSpec[] s_dotNetFrameworkSpecs =
+        private static readonly DotNetFrameworkSpec[] DotNetFrameworkSpecs =
```
Comment (ladipro) [inline src/Shared/FrameworkLocationHelper.cs:148]: Can you please make this one and `VisualStudioSpecs` lazy as well? Likely by just refactoring them into private static methods instead of fields. The methods would be called from the `Lazy` lambdas you have added.

**Review by ladipro — COMMENTED** [review]
Thank you for making the change. I left one comment which I feel is important. Please re-request review from @Forgind after addressing.

File: src/Shared/FrameworkLocationHelper.cs (hunk 13 above)
Comment (elachlan) [reply src/Shared/FrameworkLocationHelper.cs:148]: This code relies on the VisualStudioSpecs being an array so it can find the previous version by index. So VisualStudioSpecs can't be refactored into a static method. But it can be lazily loaded.
https://github.com/dotnet/msbuild/blob/daa103ebba42cb56ba18caefa74399a848905d3d/src/Shared/FrameworkLocationHelper.cs#L1472-L1484

File: src/Shared/FrameworkLocationHelper.cs [hunk 14]
```diff
@@ -72,7 +72,7 @@ internal static class FrameworkLocationHelper
         internal static readonly Version visualStudioVersion170 = new Version(17, 0);
 
         // keep this up-to-date; always point to the latest visual studio version.
-        internal static readonly Version visualStudioVersionLatest = visualStudioVersion160;
+        internal static readonly Version visualStudioVersionLatest = visualStudioVersion170;
```
Comment (ladipro) [inline src/Shared/FrameworkLocationHelper.cs:75]: Nice catch!

**Review by ladipro — APPROVED** [review]

File: src/MSBuild/XMake.cs [hunk 15]
```diff
@@ -127,7 +127,9 @@ public enum ExitType
...
         static MSBuildApp()
+#pragma warning restore CA1810 // Initialize reference type static fields inline
```
Comment (Forgind) [inline src/MSBuild/XMake.cs:132]: nit: I think I'd slightly prefer putting the restore below the end of this function, but I don't care too much.

**Review by Forgind — APPROVED** [review]
Looks good to me!
